bootstore/src/db/mod.rs

```diff
@@ -76,41 +67,95 @@ impl Db {
         // Create tables
         c.batch_execute(&schema)?;
 
-        Ok(Db { log, conn: c })
+        Ok((Db { log, path: path.to_string() }, c))
+    }
+
+    pub fn new_connection(&self) -> Result<SqliteConnection, Error> {
+        info!(
+            self.log,
+            "Creating a new connection to database {:?}", self.path
+        );
+        SqliteConnection::establish(&self.path).map_err(|err| Error::DbOpen {
+            path: self.path.clone(),
+            err: err.to_string(),
+        })
     }
 
+    /// Write an uncommitted `KeyShare` into the database.
+    ///
+    /// This command is idempotent.
+    ///
+    /// The rules for inserting a KeyShare are:
+    ///   1. A KeyShare for the given epoch does not exist unless it is identical
+    ///   2. A KeyShare for a later epoch does not exist
+    ///
+    /// Calling this method with an epoch of 0 is a programmer error so we
+    /// assert.
     pub fn prepare_share(
-        &mut self,
+        &self,
+        conn: &mut SqliteConnection,
+        rack_uuid: &Uuid,
         epoch: i32,
```

Comment:
Can epochs be negative?

Reply:
NOOOOOOOOOO. However, the type of INTEGER in sqlite is an i32, and even BIGINT is just i64. Rather than forcing lossy conversion I kept the types the same. I'm definitely open to changing this although it shouldn't be much of a problem here since we only start with epoch 0 and reject updates for earlier epochs.

Reply:
I'm tempted to suggest making `epoch` a `u32` in all the public APIs of this crate; the fact that SQLite forces us into i32 is bleeding out here. But if that's painful I don't think it's a huge deal given the checks.

Reply:
I'll think about it :)

Reply:
I'm not actually sure it's better to use a u32 given the constraints. I'm gonna leave it for now just to get this merged, but will keep it in mind.